### PR TITLE
clean log and limit log levels by change .env file

### PR DESCRIPTION
**Pull request by henryzbw**
Use `loglevel` and `dotenv` to set loglevel

### CODE REVIEW

File: .npmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+loglevel = "silent"
```
Comment (frankobe) [inline .npmrc]: newline at EOF (End-of-File)

File: src/index.js [hunk 2]
```diff
@@ -1,5 +1,10 @@
 require("babel-core/register");
 require('babel-polyfill');
+require('dotenv').config();
```
Comment (frankobe) [inline src/index.js]: can we hv loglevel `error` for `npm build` and `debug` for the `npm start`

File: src/index.js [hunk 3]
```diff
@@ -1,5 +1,10 @@
...
+
+const loglvl = process.env.loglvl || 'silent';
```
Comment (frankobe) [inline src/index.js]: where is the `process.env.loglvl` configured?

File: .gitignore [hunk 4]
```diff
@@ -131,3 +129,7 @@ $RECYCLE.BIN/
 *.db
 
 build/*.*
+
+
+# log fiel
```
Comment (dwalintukan) [inline .gitignore]: file

File: src/index.js [hunk 5]
```diff
@@ -1,5 +1,39 @@
...
+
+var fs = require('fs');
+var dir = './logs';
+
+if (!fs.existsSync(dir)){
+    fs.mkdirSync(dir);
+}
+
+const loglvl = process.env.loglvl || 'info';
+var winston = require('winston');
```
Comment (dwalintukan) [inline src/index.js]: move requires to top of file with other imports

File: src/index.js [hunk 6]
```diff
@@ -1,5 +1,39 @@
...
+var config = winston.config;
+var logger = new (winston.Logger)({
+  transports: [
+    new (winston.transports.Console)({
+      timestamp: function() {
+        return new Date().toISOString().slice(0,19); 
+      },
+      formatter: function(options) {
+        return options.timestamp() +' '+ __filename+' ' +
```
Comment (dwalintukan) [inline src/index.js]: use string templates: 
```
`${options.timestamp()} ${__filename}...`
```

File: src/index.js [hunk 7]
```diff
@@ -1,5 +1,39 @@
...
+          config.colorize(options.level, options.level.toUpperCase()) + ' ' +
+          (options.message ? options.message : '') +
+          (options.meta && Object.keys(options.meta).length ? '\n\t'+ JSON.stringify(options.meta) : '' );
+      }
+    }),
+    new (winston.transports.File)({
+      filename: './logs/'+new Date().toISOString().slice(0,10)+'.log',
```
Comment (dwalintukan) [inline src/index.js]: use string template

File: src/sync/index.js [hunk 8]
```diff
@@ -4,6 +4,41 @@ const _ = require('lodash');
 const Qweb3 = require('qweb3').default;
 const Contract = require('qweb3').Contract;
 
+
+
+const loglvl = process.env.loglvl || 'info';
+
+var winston = require('winston');
```
Comment (dwalintukan) [inline src/sync/index.js]: move to top of file with other imports.

File: src/sync/index.js [hunk 9]
```diff
@@ -4,6 +4,41 @@ const _ = require('lodash');
...
+var winston_config = winston.config;
+var logger = new (winston.Logger)({
```
Comment (dwalintukan) [inline src/sync/index.js]: looks like this entire logic is the same for both `src/index.js` and `src/sync/index.js`. can you move into a utility method so there is no code duplication?

File: src/sync/index.js [hunk 10]
```diff
@@ -4,6 +4,41 @@ const _ = require('lodash');
...
+      }
+    }),
+    new (winston.transports.File)({
+      filename: './logs/'+new Date().toISOString().slice(0,10)+'.log', 
+      
+    })
+  ]
+});
+
+
```
Comment (dwalintukan) [inline src/sync/index.js]: 👾 👾  (too many blank spaces)

File: package.json [hunk 11]
```diff
@@ -28,6 +28,8 @@
     "apollo-server-restify": "^1.3.2",
     "bluebird": "^3.5.1",
     "dataloader": "^1.3.0",
+    "dotenv": "^5.0.0",
+    "file-system": "^2.2.2",
```
Comment (frankobe) [inline package.json]: why do we need this lib?

File: src/index.js [hunk 12]
```diff
@@ -1,5 +1,39 @@
 require("babel-core/register");
 require('babel-polyfill');
+require('dotenv').config();
+
+var fs = require('fs');
```
Comment (frankobe) [inline src/index.js]: const

File: src/index.js [hunk 13]
```diff
@@ -1,5 +1,39 @@
...
+var dir = './logs';
+
+if (!fs.existsSync(dir)){
+    fs.mkdirSync(dir);
+}
+
+const loglvl = process.env.loglvl || 'info';
+var winston = require('winston');
+var config = winston.config;
+var logger = new (winston.Logger)({
```
Comment (frankobe) [inline src/index.js]: can we move this logger to a separate file like /utils/logger.js

File: src/index.js (hunk 7 above)
Comment (frankobe) [inline src/index.js]: dunt use relative path

File: src/index.js [hunk 14]
```diff
@@ -1,5 +1,39 @@
 require("babel-core/register");
 require('babel-polyfill');
+require('dotenv').config();
+
+var fs = require('fs');
+var dir = './logs';
```
Comment (frankobe) [inline src/index.js]: dunt use relative path

File: src/index.js [hunk 15]
```diff
@@ -1,5 +1,39 @@
...
+
+if (!fs.existsSync(dir)){
+    fs.mkdirSync(dir);
+}
+
+const loglvl = process.env.loglvl || 'info';
```
Comment (frankobe) [inline src/index.js]: where is `dunt use relative path` defined?

File: src/index.js (hunk 2 above)
Comment (frankobe) [reply src/index.js]: is this one used anymore?

File: package.json (hunk 11 above)
Comment (henryzbw) [reply package.json]: I use this to check if log directory exist and if it doesn't exist, create it

File: src/index.js [hunk 16]
```diff
@@ -73,7 +82,7 @@ const openBrowser = async () => {
       });
     }
   } catch(err) {
-    console.debug('Chrome not found. Launching default browser.');
+    log.debug('Chrome not found. Launching default browser.');
```
Comment (frankobe) [inline src/index.js]: `log` or `logger`?

File: src/utils/logger.js [hunk 17]
```diff
@@ -0,0 +1,23 @@
+require('dotenv').config();
+var winston = require('winston');
```
Comment (frankobe) [inline src/utils/logger.js]: const

File: src/utils/logger.js [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+
+var config = winston.config;
+var logger = new (winston.Logger)({
+  transports: [
+    new (winston.transports.Console)({
+      timestamp: function() {
+        return new Date().toISOString().slice(0,19);
```
Comment (frankobe) [inline src/utils/logger.js]: use time formatter like `YYYY-MM-DD`

File: src/utils/logger.js [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+    new (winston.transports.Console)({
+      timestamp: function() {
+        return new Date().toISOString().slice(0,19); 
+      },
+      formatter: function(options) {
+        return `${options.timestamp()} ${__filename} ${config.colorize(options.level, options.level.toUpperCase())} ${(options.message ? options.message : '')} ${(options.meta && Object.keys(options.meta).length ? '\n\t'+ JSON.stringify(options.meta) : '' )}`;
+      }
+    }),
+    new (winston.transports.File)({
+      filename: `${__dirname}/../logs/${new Date().toISOString().slice(0,10)}.log'`,
```
Comment (frankobe) [inline src/utils/logger.js]: use time formatter like `YYYY-MM-DD`

File: src/index.js [hunk 20]
```diff
@@ -26,15 +29,21 @@ const cors = corsMiddleware({
   origins: ['*'],
 });
 
+const dir = `${__dirname}/logs`;
```
Comment (frankobe) [inline src/index.js]: include this in config file

File: src/index.js [hunk 21]
```diff
@@ -83,21 +92,24 @@ const qtumdPath = path.dirname(process.argv[0])+'/qtumd';
...
 qtumprocess.stdout.on('data', (data) => {
-  console.log(`stdout: ${data}`);
+  logger.debug(`stdout: ${data}`);
 });
 
 qtumprocess.stderr.on('data', (data) => {
-  console.log(`qtum node cant start with error: ${data}`);
-  process.exit();
+  logger.error(`qtum node cant start with error: ${data}`);
+  setTimeout(() => {
```
Comment (frankobe) [inline src/index.js]: add comment on why delay here

File: src/sync/index.js [hunk 22]
```diff
@@ -242,7 +243,7 @@ async function syncCentralizedOracleCreated(db, startBlock, endBlock, removeHexP
             await db.Oracles.insert(centralOracle);
             resolve();
           } catch (err) {
-            console.error(`ERROR: ${err.message}`);
+            logger.error(`ERROR: ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove `ERROR `

File: src/sync/index.js [hunk 23]
```diff
@@ -217,13 +218,13 @@ async function syncCentralizedOracleCreated(db, startBlock, endBlock, removeHexP
       startBlock, endBlock, Contracts.EventFactory.address,
       [Contracts.OracleFactory.CentralizedOracleCreated], Contracts, removeHexPrefix,
     );
-    console.log('searchlog CentralizedOracleCreated');
+    logger.debug('searchlog CentralizedOracleCreated');
   } catch (err) {
-    console.error(`ERROR: ${err.message}`);
+    logger.error(`ERROR: ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 24]
```diff
@@ -262,13 +263,13 @@ async function syncDecentralizedOracleCreated(db, startBlock, endBlock, removeHe
       startBlock, endBlock, [], Contracts.OracleFactory.DecentralizedOracleCreated,
       Contracts, removeHexPrefix,
     );
-    console.log('searchlog DecentralizedOracleCreated');
+    logger.debug('searchlog DecentralizedOracleCreated');
   } catch (err) {
-    console.error(`ERROR: ${err.message}`);
+    logger.error(`ERROR: ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 25]
```diff
@@ -305,13 +306,13 @@ async function syncOracleResultVoted(db, startBlock, endBlock, removeHexPrefix,
       startBlock, endBlock, [], Contracts.CentralizedOracle.OracleResultVoted,
       Contracts, removeHexPrefix,
     );
-    console.log('searchlog OracleResultVoted');
+    logger.debug('searchlog OracleResultVoted');
   } catch (err) {
-    console.error(`ERROR: ${err.message}`);
+    logger.error(`ERROR: ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 26]
```diff
@@ -327,7 +328,7 @@ async function syncOracleResultVoted(db, startBlock, endBlock, removeHexPrefix,
             await db.Votes.insert(vote);
             resolve();
           } catch (err) {
-            console.error(`ERROR: ${err.message}`);
+            logger.error(`ERROR: ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 27]
```diff
@@ -286,7 +287,7 @@ async function syncDecentralizedOracleCreated(db, startBlock, endBlock, removeHe
             await db.Oracles.insert(decentralOracle);
             resolve();
           } catch (err) {
-            console.error(`ERROR: ${err.message}`);
+            logger.error(`ERROR: ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 28]
```diff
@@ -347,13 +348,13 @@ async function syncOracleResultSet(db, startBlock, endBlock, removeHexPrefix, or
       startBlock, endBlock, [], Contracts.CentralizedOracle.OracleResultSet, Contracts,
       removeHexPrefix,
     );
-    console.log('searchlog OracleResultSet');
+    logger.debug('searchlog OracleResultSet');
   } catch (err) {
-    console.error(`ERROR: ${err.message}`);
+    logger.error(`ERROR: ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 29]
```diff
@@ -371,7 +372,7 @@ async function syncOracleResultSet(db, startBlock, endBlock, removeHexPrefix, or
             );
             resolve();
           } catch (err) {
-            console.error(`ERROR: ${err.message}`);
+            logger.error(`ERROR: ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 30]
```diff
@@ -391,13 +392,13 @@ async function syncFinalResultSet(db, startBlock, endBlock, removeHexPrefix, top
       startBlock, endBlock, [], Contracts.TopicEvent.FinalResultSet, Contracts,
       removeHexPrefix,
     );
-    console.log('searchlog FinalResultSet');
+    logger.debug('searchlog FinalResultSet');
   } catch (err) {
-    console.error(`ERROR: ${err.message}`);
+    logger.error(`ERROR: ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 31]
```diff
@@ -420,7 +421,7 @@ async function syncFinalResultSet(db, startBlock, endBlock, removeHexPrefix, top
             );
             resolve();
           } catch (err) {
-            console.error(`ERROR: ${err.message}`);
+            logger.error(`ERROR: ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 32]
```diff
@@ -441,9 +442,9 @@ async function updateOraclesPassedEndBlock(currentBlockChainHeight, db) {
       { $set: { status: 'WAITRESULT' } },
       { multi: true },
     );
-    console.log('Updated Oracles Passed EndBlock');
+    logger.debug('Updated Oracles Passed EndBlock');
   } catch (err) {
-    console.error(`ERROR: updateOraclesPassedEndBlock ${err.message}`);
+    logger.error(`ERROR: updateOraclesPassedEndBlock ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 33]
```diff
@@ -454,9 +455,9 @@ async function updateCentralizedOraclesPassedResultSetEndBlock(currentBlockChain
       { resultSetEndBlock: { $lt: currentBlockChainHeight }, token: 'QTUM', status: 'WAITRESULT' },
       { $set: { status: 'OPENRESULTSET' } }, { multi: true },
     );
-    console.log('Updated COracles Passed ResultSetEndBlock');
+    logger.debug('Updated COracles Passed ResultSetEndBlock');
   } catch (err) {
-    console.error(`ERROR: updateCentralizedOraclesPassedResultSetEndBlock ${err.message}`);
+    logger.error(`ERROR: updateCentralizedOraclesPassedResultSetEndBlock ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 34]
```diff
@@ -465,11 +466,11 @@ async function updateOracleBalance(oracleAddress, topicSet, db) {
   try {
     oracle = await db.Oracles.findOne({ _id: oracleAddress });
     if (!oracle) {
-      console.error(`ERROR: find 0 oracle ${oracleAddress} in db to update`);
+      logger.error(`ERROR: find 0 oracle ${oracleAddress} in db to update`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 35]
```diff
@@ -465,11 +466,11 @@ async function updateOracleBalance(oracleAddress, topicSet, db) {
...
       return;
     }
   } catch (err) {
-    console.error(`ERROR: update oracle ${oracleAddress} in db, ${err.message}`);
+    logger.error(`ERROR: update oracle ${oracleAddress} in db, ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 36]
```diff
@@ -482,7 +483,7 @@ async function updateOracleBalance(oracleAddress, topicSet, db) {
     try {
       value = await contract.call('getTotalBets', { methodArgs: [], senderAddress });
     } catch (err) {
-      console.error(`ERROR: getTotalBets for oracle ${oracleAddress}, ${err.message}`);
+      logger.error(`ERROR: getTotalBets for oracle ${oracleAddress}, ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 37]
```diff
@@ -491,7 +492,7 @@ async function updateOracleBalance(oracleAddress, topicSet, db) {
     try {
       value = await contract.call('getTotalVotes', { methodArgs: [], senderAddress });
     } catch (err) {
-      console.error(`ERROR: getTotalVotes for oracle ${oracleAddress}, ${err.message}`);
+      logger.error(`ERROR: getTotalVotes for oracle ${oracleAddress}, ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 38]
```diff
@@ -500,9 +501,9 @@ async function updateOracleBalance(oracleAddress, topicSet, db) {
 
   try {
     await db.Oracles.update({ _id: oracleAddress }, { $set: { amounts: balances } });
-    console.log(`Update oracle ${oracleAddress} amounts ${balances}`);
+    logger.debug(`Update oracle ${oracleAddress} amounts ${balances}`);
   } catch (err) {
-    console.error(`ERROR: update oracle ${oracleAddress}, ${err.message}`);
+    logger.error(`ERROR: update oracle ${oracleAddress}, ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 39]
```diff
@@ -511,11 +512,11 @@ async function updateTopicBalance(topicAddress, db) {
   try {
     topic = await db.Topics.findOne({ _id: topicAddress });
     if (!topic) {
-      console.error(`ERROR: find 0 topic ${topicAddress} in db to update`);
+      logger.error(`ERROR: find 0 topic ${topicAddress} in db to update`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 40]
```diff
@@ -511,11 +512,11 @@ async function updateTopicBalance(topicAddress, db) {
...
       return;
     }
   } catch (err) {
-    console.error(`ERROR: find topic ${topicAddress} in db, ${err.message}`);
+    logger.error(`ERROR: find topic ${topicAddress} in db, ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 41]
```diff
@@ -528,7 +529,7 @@ async function updateTopicBalance(topicAddress, db) {
     totalBetsValue = await getTotalBetsPromise;
     totalVotesValue = await getTotalVotesPromise;
   } catch (err) {
-    console.error(`ERROR: getTotalBets for topic ${topicAddress}, ${err.message}`);
+    logger.error(`ERROR: getTotalBets for topic ${topicAddress}, ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/sync/index.js [hunk 42]
```diff
@@ -541,9 +542,9 @@ async function updateTopicBalance(topicAddress, db) {
       { _id: topicAddress },
       { $set: { qtumAmount: totalBetsBalances, botAmount: totalVotesBalances } },
     );
-    console.log(`Update topic ${topicAddress} qtumAmount ${totalBetsBalances} botAmount ${totalVotesBalances}`);
+    logger.debug(`Update topic ${topicAddress} qtumAmount ${totalBetsBalances} botAmount ${totalVotesBalances}`);
   } catch (err) {
-    console.error(`ERROR: update topic ${topicAddress} in db, ${err.message}`);
+    logger.error(`ERROR: update topic ${topicAddress} in db, ${err.message}`);
```
Comment (frankobe) [inline src/sync/index.js]: remove ERROR

File: src/utils/logger.js [hunk 43]
```diff
@@ -0,0 +1,32 @@
...
+      },
+      json: false,
+    })
+  ],
+  exitOnError: false
+});
+const loglvl = process.env.loglvl || 'info';
+logger.level = loglvl;
+
+module.exports = logger;
```
Comment (frankobe) [inline src/utils/logger.js]: newline at EOF

File: src/utils/logger.js [hunk 44]
```diff
@@ -0,0 +1,32 @@
...
+      }, 
+      formatter: function(options) {
+        return `${options.timestamp()} ${__filename} ${config.colorize(options.level, options.level.toUpperCase())} ${(options.message ? options.message : '')} ${(options.meta && Object.keys(options.meta).length ? '\n\t'+ JSON.stringify(options.meta) : '' )}`;
+      },
+      json: false,
+    })
+  ],
+  exitOnError: false
+});
+const loglvl = process.env.loglvl || 'info';
```
Comment (frankobe) [inline src/utils/logger.js]: 1. `info` should read from config file
2. where is process.env.loglvl defined?

File: src/utils/logger.js [hunk 45]
```diff
@@ -0,0 +1,32 @@
+require('dotenv').config();
```
Comment (frankobe) [inline src/utils/logger.js]: what does this line do?

File: src/utils/logger.js [hunk 46]
```diff
@@ -0,0 +1,32 @@
...
+const moment = require('moment');
+const winston = require('winston');
+
+var config = winston.config;
```
Comment (frankobe) [inline src/utils/logger.js]: what is defined in winston.config?

File: src/utils/logger.js (hunk 45 above)
Comment (henryzbw) [reply src/utils/logger.js]: This line is to require and configure dotenv. so I can get loglvl from .env file

File: src/utils/logger.js (hunk 46 above)
Comment (henryzbw) [reply src/utils/logger.js]: use winston.config to set the colorized console log

File: src/utils/logger.js (hunk 44 above)
Comment (henryzbw) [reply src/utils/logger.js]: process.env.loglvl is defined in the .env file

**Pull request merged**